### PR TITLE
test: Fixing kvbm tests/ adding concurrency test

### PR DESCRIPTION
**Pull request by oandreeva-nv**
#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx

## Release Notes

* **Tests**
  * Added concurrency stress test markers for improved test categorization and filtering
  * Enhanced test infrastructure with runtime capability detection for TensorRTLLM and vLLM availability
  * Refactored determinism testing workflow with new load-based test scenarios

* **Chores**
  * Reduced log verbosity in block manager operations

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by oandreeva-nv** [comment]
/ok to test [137edf1](https://github.com/ai-dynamo/dynamo/pull/5474/commits/137edf12e44b30b3b09935d57a8b443130c98263)

**Comment by oandreeva-nv** [comment]
/ok to test [137edf1](https://github.com/ai-dynamo/dynamo/pull/5474/commits/137edf12e44b30b3b09935d57a8b443130c98263)

**Comment by oandreeva-nv** [comment]
/ok to test [826847e](https://github.com/ai-dynamo/dynamo/pull/5474/commits/826847ef689bd1876157f42740ea510fd62c7f55)

File: tests/kvbm_integration/spanish-prompt.txt [hunk 1]
```diff
@@ -0,0 +1 @@
... de ejecución y decisiones técnicas. No interviene, salvo para recordar una norma general o ajustar el ritmo del aula cuando lo considera necesario.\n\nAl finalizar el tiempo estipulado, Divus da la señal de cierre. Recoge los materiales con precisión, verificando que todo esté completo y correctamente identificado. Luego se coloca al frente del aula, cruzando los brazos mientras observa al grupo.\n\n\"La evaluación ha concluido. Los resultados serán analizados con detenimiento. No esperen comentarios inmediatos; prefiero revisar cada procedimiento con calma.\"\n\nSu mirada se detiene brevemente en ti, sin mostrar emoción evidente, pero registrando mentalmente el desempeño observado.\n\n\"Aquellos que demuestren consistencia y criterio recibirán instrucciones adicionales en los próximos días. El aprendizaje no termina con una prueba.\"\n\nDivus abandona el aula con paso firme, dejando tras de sí un ambiente de reflexión y expectativa. Para él, cada examen es una herramienta más para pulir talento, y ya ha comenzado a planificar los siguientes pasos del curso.\n\nNo escribas como {{user}} ni asumas su reacción o respuesta. Espera la respuesta de {{user}} antes de continuar.\n[/INST]
```
[the remark this answers is not in the document]
Comment (oandreeva-nv) [reply tests/kvbm_integration/spanish-prompt.txt]: resolved

File: tests/kvbm_integration/common.py [hunk 2]
```diff
@@ -792,6 +624,325 @@ def base_test_determinism_with_cache_reset(
...
+                        f"  Waiting... ({elapsed}s elapsed, metrics check failed: {e})"
+                    )
+
+            if not benchmark_started:
+                print(f" Warning: No benchmark activity detected after {max_wait}s")
+                print(" Test will continue anyway, but may not reproduce the bug")
+            else:
+                # Give a bit more time for benchmark to ramp up
+                print("Waiting additional 10s for benchmark to fully ramp up...")
+                time.sleep(10)
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/common.py]: I am confused about this section. If we produce an exception on the metrics check, shouldn't we fail since we can't determine the metrics for kvbm?

File: tests/kvbm_integration/common.py [hunk 3]
```diff
@@ -792,6 +624,325 @@ def base_test_determinism_with_cache_reset(
...
+                            f" Waiting... ({elapsed}s elapsed, no offload activity yet)"
+                        )
+                except Exception as e:
+                    print(
+                        f"  Waiting... ({elapsed}s elapsed, metrics check failed: {e})"
+                    )
+
+            if not benchmark_started:
+                print(f" Warning: No benchmark activity detected after {max_wait}s")
+                print(" Test will continue anyway, but may not reproduce the bug")
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/common.py]: This seems ambiguous that we are continuing a test even though a criteria which we are expecting isn't available?

File: tests/kvbm_integration/common.py [hunk 4]
```diff
@@ -792,6 +624,325 @@ def base_test_determinism_with_cache_reset(
...
+                    )
+                    responses.append(response)
+                    print(f"Response: {response}")
+
+                    # Check for obvious corruption patterns (fail immediately)
+                    corruption_detected = False
+                    corruption_reason = []
+
+                    # Check for excessive repetition of short patterns (like {{char}}{{char}})
+                    if response.count("{{") > 20 or response.count("}}") > 20:
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/common.py]: I get that we are wanting to alert ourselves to corruption, but this part of the code might be better encapsulated into it's own function since the test, as its name implies, is testing determinism. This section of the code then becomes a one line comment with a function call after it to not distract the maintainer when debugging the test. 

I might argue that the corruption test is itself a unique test, I don't really have a strong opinion on this.

File: tests/kvbm_integration/common.py [hunk 5]
```diff
@@ -792,6 +624,325 @@ def base_test_determinism_with_cache_reset(
...
+                print("TEST PASSED - ALL RESPONSES DETERMINISTIC")
+                print(f"{'='*70}")
+                print(f"All {num_requests} responses matched the baseline!")
+
+            # Check final onboard stats
+            print(f"\n{'='*70}")
+            print("FINAL KVBM STATS")
+            print(f"{'='*70}")
+            try:
+                final_metrics = fetch_kvbm_metrics()
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/common.py]: More to the point above. If we didn't fail when we got a metrics exception at the beginning of the test, then we would run the test, possibly pass the actual criteria, and then fail here.

File: tests/kvbm_integration/common.py [hunk 6]
```diff
@@ -792,6 +624,325 @@ def base_test_determinism_with_cache_reset(
...
+                print(f"Final offload:   {final_offload} blocks")
+                print(f"Total offloaded: {offload_delta} blocks")
+                print(f"Total onboarded: {final_onboard} blocks")
+
+                if offload_delta > 0:
+                    print(
+                        f"\n KVBM offload activity detected: {offload_delta} blocks offloaded"
+                    )
+                else:
+                    print("\n WARNING: No offload activity detected during test")
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/common.py]: I thought one of the core functions of this test is for determinism when we offload/onload. blocks. If we didn't off load/onload any blocks, then this should fail the test and alert the maintainer to change the configuration.

File: tests/kvbm_integration/common.py [hunk 7]
```diff
@@ -792,6 +624,325 @@ def base_test_determinism_with_cache_reset(
...
+
+                if final_onboard > 0:
+                    print(
+                        f" KVBM onboard activity detected: {final_onboard} blocks onboarded"
+                    )
+                else:
+                    print(" WARNING: No onboard activity detected during test")
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/common.py]: I guess the same as [this comment](https://github.com/ai-dynamo/dynamo/pull/5474/changes#r2700265310) but slightly different

File: tests/kvbm_integration/test_cuda_graph.py [hunk 8]
```diff
@@ -187,6 +205,7 @@ def test_kvbm_without_cuda_graph_enabled(request, runtime_services):
 @pytest.mark.slow
 @pytest.mark.nightly
 @pytest.mark.gpu_1
+@pytest.mark.skipif(not HAS_TRTLLM, reason="requires tensorrt_llm")
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/test_cuda_graph.py:198]: nice!

File: tests/kvbm_integration/test_determinism_agg.py [hunk 9]
```diff
@@ -17,24 +17,44 @@
...
+    """Find a free port by binding to port 0 and letting the OS assign one."""
+
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
+        s.listen(1)
+        port = s.getsockname()[1]
+    return port
+
+
+HAS_VLLM_BENCH = _check_command_available("vllm")
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/test_determinism_agg.py]: Is this the same as how we check for trtllm?
```
def
    if importlib.util.find_spec(module_name) is None:
        return False
    try:
        importlib.import_module(module_name)
        return True
    except ImportError:
        return False
```
 If yes, my nit comment would be to keep this type of detection consistent throughout the test framework. i.e. checking the installation of the module rather than using the shutil

File: tests/kvbm_integration/test_determinism_agg.py [hunk 10]
```diff
@@ -379,197 +445,34 @@ def test_determinism_agg_with_cache_reset(
     @pytest.mark.parametrize(
         "llm_server",
         [
-            {"cpu_blocks": int(os.environ.get("KVBM_CPU_BLOCKS", "20000"))},
+            {"cpu_blocks": 20000, "gpu_blocks": 5000},
```
Comment (nv-kmcgill53) [inline tests/kvbm_integration/test_determinism_agg.py]: For my own understanding, why does this test get static values but `test_determinism_agg_with_cache_reset` does not?

File: tests/kvbm_integration/common.py (hunk 3 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: good point, I think I was debugging something and left it here, will fail

File: tests/kvbm_integration/common.py (hunk 2 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: vllm bench takes some time to start up, prepare requests and start sending it. I'll add a fail fast mechanism earlier, but this loop helps to wait for vllm bench to start up

File: tests/kvbm_integration/common.py (hunk 4 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: good point, I'll extract it. corruption of such kind just helps to fail the test fast and not wait till the end

File: tests/kvbm_integration/common.py (hunk 5 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: ditto. will add fail fast thing earlier

File: tests/kvbm_integration/test_determinism_agg.py (hunk 10 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/test_determinism_agg.py]: I can bring dynamic in dynamic. This configuration is known to repro the bug.

File: tests/kvbm_integration/common.py [hunk 11]
```diff
@@ -792,6 +624,325 @@ def base_test_determinism_with_cache_reset(
...
+            responses = []
+            baseline_response = None  # Will be set from first response
+            mismatches = []  # Track which requests had non-deterministic responses
+
+            for i in range(num_requests):
+                print(f"\n--- Request {i+1}/{num_requests} ---")
+
+                try:
+                    response = tester.make_request(
+                        prompt, max_tokens=max_tokens, temperature=0.7, seed=42
```
Comment (furionw) [inline tests/kvbm_integration/common.py]: i get that we want a fixed seed for determinism. 

i am not sure if we need to set temperature explicitly, or we can just leave it to default value 0. if default is fine, should we opt for simplicity?

**Review by furionw — COMMENTED** [review]
Thanks for inviting me to review! I learned a lot from this PR and your discussion.

Besides what Kyle mentioned, I don't have much value to add as of today.

File: tests/kvbm_integration/common.py (hunk 11 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: agree, will change it to 0/keep default 0

Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: keeping default 0

File: tests/kvbm_integration/test_determinism_agg.py (hunk 9 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/test_determinism_agg.py]: unified checks into one: `check_module_available`

File: tests/kvbm_integration/common.py (hunk 2 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: I've updated logic to fail if no offloading detected, will resolve related comments

File: tests/kvbm_integration/common.py (hunk 4 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py]: removed this complicated logic

File: tests/kvbm_integration/common.py [hunk 12]
```diff
@@ -663,6 +602,446 @@ def __init__(self):
...
+
+                except Exception as e:
+                    print(f"Request failed: {e}")
+                    responses.append(None)
+                    mismatches.append({"request_num": i + 1, "error": str(e)})
+
+                # Wait before next request
+                if i < num_requests - 1:
+                    print(f"Waiting {delay_seconds}s...")
+                    time.sleep(delay_seconds)
```
Comment (furionw) [inline tests/kvbm_integration/common.py:1024]: how about adding reasons why we need to wait here

(can you educate me why we need to wait 30 seconds between requests?

File: tests/kvbm_integration/common.py [hunk 13]
```diff
@@ -663,6 +602,446 @@ def __init__(self):
...
+
+            semantic_success_rate = (semantic_matches / num_requests) * 100
+            min_success_rate = 80.0
+
+            print(f"\n{'='*70}")
+            print(f"SEMANTIC SUCCESS RATE: {semantic_success_rate:.1f}%")
+            print(f"{'='*70}")
+            print(f"Failed requests: {[m['request_num'] for m in mismatches]}")
+
+            if semantic_success_rate < min_success_rate:
```
Comment (furionw) [inline tests/kvbm_integration/common.py:833]: correct me if i am wrong -- 

prior to this PR, we're asserting 100% exact match rate. This PR seems to assert 80% semantic (quasi-) match. are we relaxing the condition? if so, why?

asking because if we're to build a deterministic tests, we might want avoid avoid tuning threshold while tests can stay stable most of the time (if not all)


```
success_rate = deterministic_count / total_compared if total_compared > 0 else 0

...

assert (
    success_rate == 1.0
), f"Determinism failed: {deterministic_count}/{total_compared} prompts deterministic"
```

Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py:833]: yes, so there's still a numerical instability possible and what we've been observing for a while. The numerical instability is also observed on instances when you send the request, kv caches are computed and offloaded, you clear the G1 cache and send same request again, and the response will differ from the original one. This is also observed on vllm w/o kvbm.

For this case, I don't want to reduce drastically the number of output tokens, I found out that on A100 (used in ci) for output token length 80 its 100% deterministic, but for later tokens, they start to flip a bit. Thus, I've decided to add similarity check.

Potentially, one can look at this as at that position multiple tokens have nearly identical logits (e.g., logit_A = 15.328475, logit_B = 15.328472). With temp=0, argmax should pick one deterministically, but at that position even tiny differences (e.g., 1e-7) can flip the argmax when logits are this close.

File: tests/kvbm_integration/common.py (hunk 12 above)
Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py:1024]: inspired by the reproducer shared for this issue. Basically while the `vllm bench` generates the concurrent load, we send the control request periodically and record the response

Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py:1024]: after we check the responses. They should be deterministic, but also some numeric instability is still possible as explained in the other comment

Comment (furionw) [reply tests/kvbm_integration/common.py:1024]: sg. shall we link https://github.com/ai-dynamo/dynamo/issues/5068 in the comments to so that people understand we put wait here because it initially helped us reproduce the issue?

Comment (oandreeva-nv) [reply tests/kvbm_integration/common.py:1024]: I think it's fine for now

File: tests/kvbm_integration/common.py (hunk 13 above)
Comment (nv-kmcgill53) [reply tests/kvbm_integration/common.py:833]: >  (e.g., logit_A = 15.328475, logit_B = 15.328472).

This might be classic floating point error and using a float compare might be better.
```Python
EPSILON = 0.0005
def bool is_equal(a: float, b: float) {
    return abs(a-b) < EPSILON;
}
```

**Review by nv-kmcgill53 — APPROVED** [review]

**Review by ryanolson — APPROVED** [review]

**Comment by dagil-nvidia** [comment]
Auto-linked to [DIS-1276](https://linear.app/nvidia/issue/DIS-1276)

**Comment by dagil-nvidia** [comment]
Auto-linked to [DYN-1789](https://linear.app/nvidia/issue/DYN-1789)
